components/brave_new_tab_ui/state/gridSitesState.ts

```diff
@@ -50,11 +50,33 @@ export function gridSitesReducerSetDefaultSuperReferralTopSites (
 
 export function gridSitesReducerSetFirstRenderData (
   state: NewTab.GridSitesState,
-  topSites: chrome.topSites.MostVisitedURL[]
+  topSites: chrome.topSites.MostVisitedURL[],
+  defaultSuperReferralTopSites: NewTab.DefaultSuperReferralTopSite[]
 ): NewTab.GridSitesState {
-  const gridSitesWhitelist = getGridSitesWhitelist(topSites)
+  const topSitesWhitelisted = getTopSitesWhitelist(topSites)
+  // |state.gridSites| has lastly used sites data for NTP.
+  // Delete sites from |state.gridSites| that don't exist in topSites(history).
+  // Then, |updatedGridSites| will only store previously used sites that topSites have.
+  const updatedGridSites: NewTab.Site[] = state.gridSites.filter((gridSite: NewTab.Site) => {
+    // Pinned tab will be remained even if it's deleted from history.
+    // If unpinned, it will not be visible if it's not in topSites.
+    if (gridSite.pinnedIndex !== undefined) {
```

Comment:
Not clearing the pinned items could be seen as a leak- if user picks "clear all time", I think it should clear everything basically

Reply:
If we wanted to make the code more "redux-ish", we could create a reducer event which clears `state.gridSites`; it would need to be sent by the C++ side (API), but it could basically execute a new action which does a new `case types.NEW_TAB_CLEAR_TOPSITES:`

(see https://github.com/brave/brave-core/blob/264b6a60eef1dffb00261538dabc85f685b9de52/components/brave_new_tab_ui/reducers/new_tab_reducer.ts#L31)

Here's an example showing listeners for preference that are calling onChange handlers (executed when setting changes in C++ side):
https://github.com/brave/brave-core/blob/3d89940e2c3ea842eaccc759303f6900c7b6657e/components/brave_new_tab_ui/apiEventsToStore.ts

@petemill what do you think? 😄

Reply:
@bsclifton how would that work if there are now NTP tabs open at that time?@bsclifton how would that work if there are now NTP tabs open at that time?

Reply:
@bsclifton Ok, will delete pinned items also if they were disappeared from history.
and I think updating topsites only with reloading would make logic simple. WDYT?

Reply:
@petemill good point - that would be overly complex 😄 

Change made looks great! 😄👍